src/main/scala/com/cloudant/clouseau/IndexService.scala

```diff
@@ -932,4 +914,170 @@ object IndexService {
+    val asListOfStrings: PartialFunction[Any, List[String]] = ensureElementsType(
+      { case string: String => string },
+      { case (container, element) => throw new ParseException(container.toString + " contains non-string element " + element.toString) }
+    )
+
+    val asListofListsOfStrings: PartialFunction[Any, List[List[String]]] = ensureElementsType(
+      asListOfStrings,
+      { case (container, element) => throw new ParseException("invalid drilldown query " + container.toString) }
+    )
+
+    val extractor = asListofListsOfStrings.Extractor
+    getOption[Any](options, 'drilldown) match {
+      case Some(extractor(value)) =>
+        Some(value)
+      case None =>
+        None
+    }
+  }
+
+  /**
+   * Returns the Option[value] of a `ranges` key from given map of options if it is in correct format.
+   *
+   * @param options: Map[Symbol, Any] - map of options to extract `ranges` key
+   * @return An optional value of a `ranges` key as `Option[List[Product2[RangesName, List[Product2[RangesLabel, RangesQuery]]]]]`
+   * @throws ParseException
+   *
+   * The function expects the `ranges` value to be in a following format
+   *
+   * ```erlang
+   * -type name :: string().
+   * -type label :: string().
+   * -type query :: string().
+   * [{name(), [
+   *   {label(), query()}
+   * ]}]
+   * ```
+   *
+   * or the same in scala format
+   *
+   * ```scala
+   * type Name = String
+   * type Label = String
+   * type Query = String
+   * List[(Name, List(
+   *   (Label, Query)
+   * ))]
+   * ```
+   *
+   */
+  private[clouseau]type RangesLabel = String
+  private[clouseau]type RangesQuery = String
+  private[clouseau]type RangesName = String
+
+  private[clouseau] def getRanges(options: Map[Symbol, Any]): Option[List[Product2[RangesName, List[Product2[RangesLabel, RangesQuery]]]]] = {
+    val asQueries: PartialFunction[Any, List[Product2[RangesLabel, RangesQuery]]] = ensureElementsType(
+      { case (label: RangesLabel, query: RangesQuery) => (label, query) },
+      { case (_container, _element) => throw new ParseException("invalid ranges query") }
+    )
+
+    val asQueriesExtractor = asQueries.Extractor
+
+    val asRange: PartialFunction[Any, List[Product2[RangesName, List[Product2[RangesLabel, RangesQuery]]]]] = ensureElementsType(
+      { case (name: RangesName, asQueriesExtractor(queries)) => (name, queries) },
+      { case (container, _element) => throw new ParseException(container.toString + " is not a valid ranges query") }
+    ).orElse({ case notAList => throw new ParseException("invalid ranges query") })
+
+    val extractor = asRange.Extractor
+
+    getOption[Any](options, 'ranges) match {
+      case Some(extractor(value)) =>
+        Some(value)
+      case None =>
+        None
+    }
+  }
+
+  /**
+   * Returns the Option[List[String]] of a given option key from a map of options if each element of a list is indeed a String.
+   * It also throws ParseException if passed value is not a List[_].
```

Comment:
There is an extra redundant mention of `ParseException` (c.f. the `@throws` clause).

Reply:
addressed in a12e954